src/actions/hover.rs

```diff
@@ -0,0 +1,480 @@
+        Ok(_) => {
+            let utf8 = String::from_utf8(out);
+            if let Ok((Some(pos), lines)) = utf8.map(|lines| (lines.rfind("{"), lines)) {
+                lines[0..pos].into()
+            } else {
+                the_type
+            }
+        },
+        Err(e) => {
+            warn!("hover: object format failed: {:?}", e);
+            the_type
+        }
+    }
+}
+
+/// Formats a method or function. The original type is returned
+/// in the event of an error.
+fn format_method(ctx: &InitActionContext, the_type: String) -> String {
+    trace!("hover: format_method: {}", the_type);
+    let config = ctx.fmt_config().get_rustfmt_config().clone();
+    let method = format!("impl Dummy {{ {} {{ unimplmented!() }} }}", the_type);
+    let mut out = Vec::<u8>::with_capacity(the_type.len());
+    match format_input(FmtInput::Text(method), &config, Some(&mut out)) {
+        Ok(_) => {
+            if let Ok(mut lines) = String::from_utf8(out) {
+                if let Some(front_pos) = lines.find("{") {
+                    lines = lines[1+front_pos..].into();
+                }
+                if let Some(back_pos) = lines.rfind("{") {
+                    lines = lines[0..back_pos].into();
+                }
+                lines.lines().filter(|line| line.trim() != "").map(|line| {
+                    format!("{}\n", &line[config.tab_spaces()..])
+                }).collect()
+            } else {
+                the_type
+            }
+        },
+        Err(e) => {
+            warn!("hover: method format failed: {:?}", e);
+            the_type
+        }
+    }
+}
+
+pub fn tooltip(
+    ctx: &InitActionContext,
+    params: &TextDocumentPositionParams
+) -> Result<Vec<MarkedString>, ResponseError> {
+    let analysis = &ctx.analysis;
+
+    let hover_file_path = parse_file_path!(&params.text_document.uri, "hover")?;
+    let hover_span = ctx.convert_pos_to_span(hover_file_path, params.position);
+    let hover_span_ty = analysis.show_type(&hover_span).unwrap_or_else(|_| String::new());
+    let hover_span_def = analysis.id(&hover_span).and_then(|id| analysis.get_def(id));
+    let hover_span_docs = analysis.docs(&hover_span).unwrap_or_else(|_| String::new());
+
+    trace!("hover: span: {:?}", hover_span);
+    trace!("hover: span_def: {:?}", hover_span_def);
+    trace!("hover: span_ty: {:?}", hover_span_ty);
+
+    let doc_url = analysis.doc_url(&hover_span).ok();
+    
+    let mut contents = vec![];
+
+    let racer_fallback = |contents: &mut Vec<MarkedString>| {
+        if let Some((racer_docs, racer_ty)) = racer_docs(&ctx, &hover_span, None) {
+            let docs = process_docs(&racer_docs);
+            let docs = if docs.trim().is_empty() { hover_span_docs } else { docs };
+            let ty = racer_ty.unwrap_or(hover_span_ty);
+            let ty = ty.trim();
+            if !ty.is_empty() {
+                contents.push(MarkedString::from_language_code("rust".into(), format!("{}", ty)));
```

Comment:
Minor, but the `format!("{}", foo_str)`  usages could just be `foo_str.into()` / `to_owned()`

Reply:
I think at one point there was newline padding in here as well. I agree with your above comments that adding extra newlines isn't so wonderful. It is definitely to fix issues with the way it's rendered in vscode. I almost left a comment in the source. 

I have been trying to figure out why rls-vscode does not include line separators (between each element of the `HoverContents::Array(Vec<MarkedString>)` data. I've looked at the hover providers for both the Go extension and the other [rust extension](https://github.com/editor-rs/vscode-rust/blob/master/src/components/completion/completion_manager.ts#L278) and I haven't been able to figure out why rls-vscode is behaving differently. Both of those extensions have explicit hoverProviders in typescript, but as far as I can tell the logic is the same. 

Note the thin black lines between the defintion, doc url, and the docs (the screen shot is using the old [rust extension](https://github.com/editor-rs/vscode-rust/blob/master/src/components/completion/completion_manager.ts#L278) ). For some reason those lines are missing with the [rls-vscode](https://github.com/rust-lang-nursery/rls-vscode) extension. 

![image](https://user-images.githubusercontent.com/221559/42572590-c185df46-84e7-11e8-8cb5-18604aa47cc1.png)

I've been trying to figure this out for a few days. Any help would be greatly appreciated!

Thanks for the review. I'll take a look at the remaining items later this evening.

Reply:
Well vscode isn't my area & atom handles this ok, _ie has the line separators_.
![box-hover](https://user-images.githubusercontent.com/2331607/42573068-006b4658-8513-11e8-9ae1-2494dc5eb1ba.png)

I'd simply raise it as a vscode issue & be fairly satisfied that rls is doing the right thing, but I appreciate you want to have it looking right there off the bat.